core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -325,13 +325,24 @@ sealed abstract class Cause[+E] extends Product with Serializable { self =>
   /**
    * Determines if the `Cause` is empty.
    */
-  final def isEmpty: Boolean =
-    (self eq Empty) || foldLeft(true) {
-      case (acc, _: Empty.type) => acc
-      case (_, Die(_, _))       => false
-      case (_, Fail(_, _))      => false
-      case (_, Interrupt(_, _)) => false
-    }
+  final def isEmpty: Boolean = {
+    @tailrec
+    def loop(cause: Cause[E], stack: List[Cause[E]]): Boolean =
```

Comment:
I almost feel it's a mistake to make all the Cause operations stack safe, because of the overhead in allocation.

But anyway, this seems a good change. 👍

Reply:
I thought about the same thing. Unfortunately I do think that users sometimes create very large nested causes, for example using `collectAllPar`. Hopefully the check for reference equality with the empty cause covers the vast majority of "happy path" cases and we are really only doing this when a failure has occurred.

One alternative would be to ensure that we construct a balanced tree so that we never reach a sufficient depth where we need to be stack safe similar to what we do with `Chunk`. This would have some complexities with the fact that `Cause` has both a `Both` and a `Then` case but it is something we could explore.